sigmastate/src/main/scala/sigmastate/Values.scala

```diff
@@ -1522,4 +1523,121 @@ object Values {
       withSegregation(DefaultHeader, prop)
   }
 
+  case class Parameter private[sigmastate](
+    name: String,
+    description: String,
+    placeholder: Int
+  ) {
+    def serializeInto(w: SigmaByteWriter): Unit = {
+      ParameterSerializer.serialize(this, w)
+    }
+  }
+
+  object Parameter {
+    /** Immutable empty IndexedSeq, can be used to save allocations in many places. */
+    val EmptySeq: IndexedSeq[Parameter] = Array.empty[Parameter]
+  }
+
+
+  /** This is alternative representation of ContractTemplate when it cannot be parsed
+    * due to `error`. */
+  case class UnparsedContractTemplate(bytes: mutable.WrappedArray[Byte], error: ValidationException)
+
+  case class ContractTemplate private[sigmastate](
+    treeVersion: Option[Byte],
+    name: String,
+    description: String,
+    constTypes: IndexedSeq[SType],
+    constValues: Option[IndexedSeq[Option[SType#WrappedType]]],
+    parameters: IndexedSeq[Parameter],
+    expressionTree: Either[UnparsedContractTemplate, SigmaPropValue],
+    private val templateBytes: Array[Byte],
+ ) {
+
+    require(constValues.isEmpty || constValues.get.size == constTypes.size,
+      s"constValues must be empty or of same length as constTypes. Got ${constValues.get.size}, expected ${constTypes.size}")
+    require(parameters.size <= constTypes.size, "number of parameters must be <= number of constants")
+
+    def this(treeVersion: Option[Byte],
+             name: String,
+             description: String,
+             constTypes: IndexedSeq[SType],
+             constValues: Option[IndexedSeq[Option[SType#WrappedType]]],
+             parameters: IndexedSeq[Parameter],
+             expressionTree: Either[UnparsedContractTemplate, SigmaPropValue]) =
+      this(
+        treeVersion, name, description, constTypes, constValues, parameters, expressionTree,
+        templateBytes = ContractTemplateSerializer.DefaultSerializer.serializeContractTemplate(
+          ContractTemplate(treeVersion, name, description, constTypes, constValues, parameters, expressionTree, null)
+        )
+      )
+
+    @inline final def isRightParsed: Boolean = expressionTree.isRight
+
+    private[sigmastate] var _bytes: Array[Byte] = templateBytes
+
+    /** Serialized bytes of this template. */
+    final def bytes: Array[Byte] = {
+      if (_bytes == null) {
+        _bytes = ContractTemplateSerializer.DefaultSerializer.serializeContractTemplate(this)
+      }
+      _bytes
+    }
+
+    def applyTemplate(paramValues: IndexedSeq[Constant[SType]]): ErgoTree = {
```

Comment:
applyTemplate should take Map[String, Constant] as argument, so parameters a substituted by name.

Reply:
Yeah, updated the code.